File: C code/secant.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

double secantMethod();

double secantMethod(double (*f)(), double x0, double x1, double tol, double maxIter)
{
    double error = 10.0 * tol;
    double iteration = 0.0;
    double f0 = f(x0);
    double f1 = f(x1);
    while(error > tol && iteration < maxIter)
    {
        double xi = x1 - (f1 * (x1 - x0)) / (f1 - f0);
        error = fabs(xi - x1);
        x0 = x1;
        x1 = xi;
        f0 = f1;
        f1 = f(xi);
        iteration = iteration + 1.0;
    }
    return x1;
}

File: Python Code/Root Finding/secant.py
import numpy as np


def f_of_x(x):
    return x * np.exp(-x)

def secantMethod(func, x0, x1, maxIter):
    iteration = 0
    tol = .00000001
    error = 10 * tol

    while (iteration < maxIter and error > tol):
        fx0 = f_of_x(x0)
        fx1 = f_of_x(x1)
        xi = x1 - (fx1 * (x1 - x0)) / (fx1 - fx0)
        error = np.abs(xi - x1)
        x0 = x1
        x1 = xi
        fx0 = fx1
        iteration += 1
    print(f"The root is found at {x1} after {iteration} iterations")

secantMethod(f_of_x, -1, 1, 25)
